### PR TITLE
promql: more Kahan summation (avg) and less incremental mean calculation (avg, avg_over_time)

### PR DESCRIPTION
**Pull request by beorn7**
See various commits for details.

Studying #14074, I noticed that the tests added for the `sum` aggregation wouldn't work for the `avg` aggregation. Also, the tests already there for `sum_over_time` (with a series `1 1e100 1 -1e100`) wouldn't work for `avg_over_time`.

Simply applying Kahan summation wouldn't help. (In fact, it was already applied to `avg_over_time`, but of course, without adding the test mentioned above.) The reason is that we always use incremental calculation of the mean value. That's a good approach if the sum of the averaged values overflow float64. But in general, it's just more costly and has way more potential of accumulating numerical errors.

This PR changes the behavior that incremental calculation of the mean is only used if the sum would overflow. It also uses Kahan summation in all cases, which should take care of errors happening when adding small numbers to large ones (which is a case where incremental calculation of the mean might help, too, but I believe Kahan summation is cheaper and better and also consistent with how we calculate sums).

### CODE REVIEW

**Comment by beorn7** [comment]
BTW, the speedup for `avg` calculation isn't significant, but it should be noted that this PR adds Kahan summation to `avg`, which has lead to a measurable slowdown for `sum` in #14074. Or in other words, the slowdown caused by the Kahan summation is approximately compensated by the speedup gained from removing incremental mean calculation.

```
name                                          old time/op    new time/op    delta
RangeQuery/expr=avg(a_hundred),steps=1-16        224µs ± 2%     216µs ± 1%  -3.39%  (p=0.000 n=10+9)
RangeQuery/expr=avg(a_hundred),steps=100-16      683µs ± 1%     686µs ± 4%    ~     (p=0.780 n=9+10)
RangeQuery/expr=avg(a_hundred),steps=1000-16    4.29ms ± 7%    4.34ms ± 2%    ~     (p=0.393 n=10+10)
```

**Comment by beorn7** [comment]
@bboreham I nominate you as the reviewer because you inspired this through #14074 .

File: promql/engine.go
Comment (bboreham) [inline promql/engine.go:2871]: Could put a `break` here and out-dent the following.

File: promql/engine.go
Comment (bboreham) [inline promql/engine.go:2882]: What sort of situation makes this useful?  We've overflowed a float64, but dividing by groupCount brings us back within range?

File: promql/functions.go
Comment (bboreham) [inline promql/functions.go:577]: Maybe rename `c` to `kahanC`.

File: promql/functions.go
Comment (bboreham) [inline promql/functions.go:584]: Why do we care about count=1 ?

**Review by bboreham — APPROVED** [review]
Looks ok, but I have a number of thoughts.

File: promql/engine.go
Comment (beorn7) [reply promql/engine.go:2882]: I have added a comment.

File: promql/engine.go
Comment (beorn7) [reply promql/engine.go:2889]: We _would_ have overflowed, but instead of actually doing it, we revert to incremental mean calculation, whose whole purpose is to avoid overflowing by "dividing early" (with some tricks). This is what we have done so far before this PR, even in (the vast majority of) cases that don't overflow.

I have tried to make the comment a bit more explanatory.

File: promql/functions.go
Comment (beorn7) [reply promql/functions.go]: Done.

File: promql/functions.go
Comment (beorn7) [reply promql/functions.go:589]: We would get division-by-zero problems below if the very first value to sum up is already ±Inf. I have added a comment to explain that. In case of the aggregators, the flow is different because we essentially deal with the first step during the initialization, which happens for all the different aggregations in the same way.

**Review by beorn7 — COMMENTED** [review]
Thanks for the review.

I have addressed all comments, but you might want to double check if it happened to your satisfaction, so I'll wait for your response before merging.

File: promql/engine.go
Comment (bboreham) [reply promql/engine.go:2889]: Right, I mean what sort of data will bring us to this case.
Perhaps 1,000 series each of whose value is 2e305 ?
And I am hinting that it may not be a very common case.

Comment (beorn7) [reply promql/engine.go:2889]: Yes, I would assume it's very rare in practice.
And that's one of the motivations behind this change.
Prior to this change, we always did incremental mean calculation, which is expensive and less accurate.
With this change, we do "normal" mean calculation, backed by Kahan summation, and only switch to incremental mean calculation in the rare cases that we need it.

**Review by bboreham — APPROVED** [review]

**Pull request merged**
